### PR TITLE
Cleaner/nicer to use command-line interface (+ help text) for the standalone player

### PR DESCRIPTION
**Pull request by PoroCYon**
Not a requirement, but definitely nice to have.

### CODE REVIEW

File: WaveSabreStandAlonePlayer/main.cpp [hunk 1]
```diff
@@ -34,20 +36,139 @@ void progressCallback(double progress, void *data)
...
+	int nthreads;
+	bool prerender;
+	bool wavwriter;
+};
+
+static void print_usage(const char* prgm)
+{
+	printf("%s: WaveSabre standalone song player/renderer\n"
+		"Usage:\n"
+		"\t%s [/p|/prerender] [/w|/wav [out.wav]] [/t|/threads 3] <input.bin>\n"
```
Comment (kusma) [inline WaveSabreStandAlonePlayer/main.cpp:52]: I would use dashes as prefixes rather than slashes, as that's not so Windows-centric. Most portable applications use dashes, as that's what the POSIX standard mandates...

File: WaveSabreStandAlonePlayer/main.cpp [hunk 2]
```diff
@@ -34,20 +36,139 @@ void progressCallback(double progress, void *data)
...
+		"Arguments:\n"
+		"\tprerender prerender the song instead of rendering while playing.\n"
+		"\twav       instead of playing back the song, write a WAV file. By default,\n"
+		"\t          the output file is called `out.wav', but another may be supplied\n"
+		"\t          immediately after this argument.\n"
+		"\tthreads   the amount of threads to use in parallel for rendering. By\n"
+		"\t          default, this number is 3.\n"
+		"\tinput.bin the input song file, exported by the WaveSabre exporter.\n"
+		, prgm, prgm);
+	exit(0);
```
Comment (kusma) [inline WaveSabreStandAlonePlayer/main.cpp:63]: Is zero the right exit-code to use here? I think usage should exit with an error, no?

File: WaveSabreStandAlonePlayer/main.cpp [hunk 3]
```diff
@@ -34,20 +36,139 @@ void progressCallback(double progress, void *data)
...
+			}
+
+			args->infile = argv[i];
+		}
+	}
+
+	if (args->infile == NULL)
+	{
+		if (args->outfile != NULL)
+		{
```
Comment (kusma) [inline WaveSabreStandAlonePlayer/main.cpp:144]: No need for two if-statements here, `if (args->infile == NULL && args->outfile != NULL)` seems more to the point.

File: WaveSabreStandAlonePlayer/main.cpp [hunk 4]
```diff
@@ -34,20 +36,139 @@ void progressCallback(double progress, void *data)
...
+			// oops, was probably meant as input file path
+			args->infile = args->outfile;
+			args->outfile = NULL;
```
Comment (kusma) [inline WaveSabreStandAlonePlayer/main.cpp:147]: Should we really make assumptions like this?

File: WaveSabreStandAlonePlayer/main.cpp [hunk 5]
```diff
@@ -34,20 +36,139 @@ void progressCallback(double progress, void *data)
...
+		}
+	}
+
+	if (args->outfile == NULL && args->wavwriter)
+	{
+		args->outfile = "out.wav";
+	}
```
Comment (kusma) [inline WaveSabreStandAlonePlayer/main.cpp:154]: I would just do this where we set args->wavwriter to true above.

File: WaveSabreStandAlonePlayer/main.cpp [hunk 6]
```diff
@@ -34,20 +36,139 @@ void progressCallback(double progress, void *data)
...
+
+	char argpfix = '/';
+
+	for (int i = 1; i < argc; ++i)
+	{
+		if (argv[i][0] == argpfix)
+		{
+			const char* as = &argv[i][1];
+
+			if (!strcmp(as, "h") || !strcmp(as, "-help") || !strcmp(as, "help"))
```
Comment (kusma) [inline WaveSabreStandAlonePlayer/main.cpp:87]: The dashed ways here makes no sense to me. Since we already check for slashes, it would allow things like "/-help" which makes very little sense to me...

Comment (kusma) [reply WaveSabreStandAlonePlayer/main.cpp:87]: I guess this code started out trying to support long arguments, so that would be "-h" and "--help", and then later changed to slashes? If so, I would just drop slashed support, as said before, then you can drop the last case here and everything becomes easier...

File: WaveSabreStandAlonePlayer/main.cpp [hunk 7]
```diff
@@ -108,7 +240,11 @@ int main(int argc, char **argv)
 			int seconds = (int)songPos % 60;
 			int hundredths = (int)(songPos * 100.0) % 100;
 			printf("\r %.1i:%.2i.%.2i", minutes, seconds, hundredths);
-
+#ifndef _MSC_VER
+			// no \n at the end of the printf above
+			// -> might only buffer, not guaranteed to actually write to stdout
+			fflush(stdout);
+#endif
```
Comment (kusma) [inline WaveSabreStandAlonePlayer/main.cpp:247]: This seems unrelated to the rest of the changes here...

File: WaveSabreStandAlonePlayer/main.cpp [hunk 8]
```diff
@@ -68,37 +192,45 @@ int main(int argc, char **argv)
...
 			printf("\n\n");
 		}
 		else
 		{
-			player = new RealtimePlayer(&song, numRenderThreads);
+			player = new RealtimePlayer(&song, args.nthreads);
 		}
 
 		printf("Realtime player activated. Press ESC to quit.\n");
+
```
Comment (kusma) [inline WaveSabreStandAlonePlayer/main.cpp:233]: Seems unrelated.

File: WaveSabreStandAlonePlayer/main.cpp (hunk 1 above)
Comment (PoroCYon) [reply WaveSabreStandAlonePlayer/main.cpp:52]: Originally (#49 ) I had either dashes or slashes depending on the target platform, but I guess I'll switch it to dashes instead.

File: WaveSabreStandAlonePlayer/main.cpp (hunk 2 above)
Comment (PoroCYon) [reply WaveSabreStandAlonePlayer/main.cpp:63]: Well this function is called both when just running the replayer without any args, and with the help flag. Common terminal tools seem to only print a "use --help for usage"-error and exit with 1 when ran without args. I guess I'll make this function just return, and the callers will call exit with the correct code.

File: WaveSabreStandAlonePlayer/main.cpp (hunk 6 above)
Comment (PoroCYon) [reply WaveSabreStandAlonePlayer/main.cpp:87]: See earlier comment, originally, the prefix char (a few lines above this code) was target-dependent.

File: WaveSabreStandAlonePlayer/main.cpp (hunk 4 above)
Comment (PoroCYon) [reply WaveSabreStandAlonePlayer/main.cpp:147]: I see no other condition that can cause this to arise, and, if this happens, it's not a good thing. But I'll handle it differently.

File: WaveSabreStandAlonePlayer/main.cpp [hunk 9]
```diff
@@ -34,20 +36,137 @@ void progressCallback(double progress, void *data)
...
+	}
+
+	args->infile = NULL;
+	args->outfile = NULL;
+	args->nthreads = 3;
+	args->prerender = false;
+	args->wavwriter = false;
 
-	const int numRenderThreads = 3;
+	const char argpfix = '-';
```
Comment (kusma) [inline WaveSabreStandAlonePlayer/main.cpp:78]: Nit: I find it *more* confusing to read "argpfix" than '-', I think :P

File: WaveSabreStandAlonePlayer/main.cpp [hunk 10]
```diff
@@ -34,20 +36,137 @@ void progressCallback(double progress, void *data)
...
+	int nthreads;
+	bool prerender;
+	bool wavwriter;
+};
+
+static void print_usage(const char* prgm)
+{
+	printf("%s: WaveSabre standalone song player/renderer\n"
+		"Usage:\n"
+		"\t%s [-p|--prerender] [-w|--wav [out.wav]] [-t|--threads 3] <input.bin>\n"
```
Comment (kusma) [inline WaveSabreStandAlonePlayer/main.cpp:52]: I don't think it's common to show the default argument value like this in usages, is it? I would just use 'n' here, I think.

File: WaveSabreStandAlonePlayer/main.cpp [hunk 11]
```diff
@@ -34,20 +36,137 @@ void progressCallback(double progress, void *data)
...
-	const int numRenderThreads = 3;
+	const char argpfix = '-';
+
+	for (int i = 1; i < argc; ++i)
+	{
+		if (argv[i][0] == argpfix)
+		{
+			const char* as = argv[i] + 1; // skip arg prefix char ('-')
+
+			if (!strcmp(as, "h") || !strcmp(as, "-help") || !strcmp(as, "help"))
```
Comment (kusma) [inline WaveSabreStandAlonePlayer/main.cpp:86]: I don't think supporting "-help" is a good idea. It's either "-h" or "--help" in most CLI interfaces. In fact, "-help" usually means the same as "-h -e -l -p".

Comment (PoroCYon) [reply WaveSabreStandAlonePlayer/main.cpp]: Fair, it was a leftover from the `/help` stuff.

It doesn't support multiple short options specified like that though, mostly because I was lazy. (I could use `getopt(3)` for that, but, that doesn't exist on Windows of course.)

Comment (kusma) [reply WaveSabreStandAlonePlayer/main.cpp]: Yeah, I think it's fine to drop support for that. I just don't think we should introduce support for long names with a single dash. Thanks for the fix :)

**Comment by kusma** [comment]
This looks great to me, nice work!

**Comment by yupferris** [comment]
I don't think this is a good idea. The player test ends up being forked very often for actual intros and exemusic prods, and thus this interface is intentionally kept minimal in order to not take up any space that we don't need to. What we currently have should be more than sufficient.
